policy/nomad/source.go

```diff
@@ -71,56 +85,70 @@ func (s *Source) MonitorIDs(ctx context.Context, req policy.MonitorIDsReq) {
 	q := &api.QueryOptions{WaitTime: 5 * time.Minute, WaitIndex: 1}
 
 	for {
+		var (
+			policies []*api.ScalingPolicyListStub
+			meta     *api.QueryMeta
+			err      error
+		)
+
+		// Perform a blocking query on the Nomad API that returns a stub list
+		// of scaling policies. The call is done in a goroutine so we can
+		// still listen for the context closing or a reload request.
+		blockingQueryCompleteCh := make(chan struct{})
+		go func() {
+			// Obtain a handler now so we can release the lock before starting
+			// the blocking query.
+			s.nomadLock.RLock()
+			scaling := s.nomad.Scaling()
+			s.nomadLock.RUnlock()
+
+			policies, meta, err = scaling.ListPolicies(q)
+			close(blockingQueryCompleteCh)
+		}()
+
 		select {
 		case <-ctx.Done():
 			s.log.Trace("stopping ID subscription")
 			return
-		default:
-			// Perform a blocking query on the Nomad API that returns a stub list
-			// of scaling policies. If we get an errors at this point, we should
-			// sleep and try again.
-			//
-			// TODO(jrasell) in the future maybe use a better method than sleep.
-			policies, meta, err := s.nomad.Scaling().ListPolicies(q)
-
-			// Return immediately if context is closed.
-			if ctx.Err() != nil {
-				s.log.Trace("stopping ID subscription")
-				return
-			}
+		case <-s.reloadCh:
+			s.log.Trace("reloading policies")
+			continue
+		case <-blockingQueryCompleteCh:
+		}
 
-			if err != nil {
-				policy.HandleSourceError(s.Name(), fmt.Errorf("failed to call the Nomad list policies API: %v", err), req.ErrCh)
-				time.Sleep(10 * time.Second)
-				continue
-			}
+		// If we get an errors at this point, we should sleep and try again.
+		// TODO(jrasell) in the future maybe use a better method than sleep.
+		if err != nil {
+			policy.HandleSourceError(s.Name(), fmt.Errorf("failed to call the Nomad list policies API: %v", err), req.ErrCh)
+			time.Sleep(10 * time.Second)
```

Comment:
while we're here, maybe this should be a select over `time.After` and `ctx.Done`, to keep an error from holding up our shutdown for 10s?